### ISSUE TITLE
Face normal reverses when loading simple obj file

### ISSUE THREAD

**Issue by humanbeingZ** [tinyobjloader/tinyobjloader#330]
This may sound like #319. However I'm not sure.

obj file:
```
v -105.342712 40.184242 -16.056709
v -105.463989 40.202003 -16.003181
v -105.564941 40.207558 -15.934708
v -105.722252 40.151146 -16.112091
v -105.610237 40.191372 -16.176643
v -105.667282 40.189800 -15.864197
v -105.751717 40.125790 -15.794304
# 7 vertices, 0 vertices normals

f 2 5 4 3
f 4 6 3
# 2 faces, 0 coords texture

# End of File
```

python3 code:
```python
import sys
import numpy as np
import tinyobjloader


def process(obj_path):
    reader = tinyobjloader.ObjReader()
    ret = reader.ParseFromFile(obj_path)

    if not ret:
        print('Failed to load: ', obj_path)
        sys.exit(-1)
    
    attrib = reader.GetAttrib()
    vertices = np.array(attrib.vertices).reshape(-1, 3)
    shape = reader.GetShapes()[0]

    indices = shape.mesh.indices
    faces = [idx.vertex_index for idx in indices]
    faces = np.array(faces).reshape(-1, 3)

    with open(f'out.off', 'w') as f:
        print('OFF', file=f)
        print(f'{len(vertices)} {len(faces)} 0', file=f)
        np.savetxt(f, vertices, fmt='%04f')
        for face in faces:
            print(3, *face, sep=' ', file=f)


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('Usage: python <obj_path>')
        sys.exit(0)

    obj_path = sys.argv[1]
    process(obj_path)

```

**Comment by humanbeingZ** [comment]
![result](https://user-images.githubusercontent.com/25080075/155736403-2c50f794-8e18-4a8e-9af3-97ef23d01fd8.png)

**Comment by syoyo** [comment]
Current tinyobjloader's built-in triangulation is not perfect. This may be solved one #319 has been fixed.

For a while, you can triangulate loaded mesh by yourself(or use other mesh triangulation library such like https://github.com/linuxlewis/tripy ) by turning `triangulation` flag off when loading .obj: https://github.com/tinyobjloader/tinyobjloader/blob/8322e00ae685ea623ab6ac5a6cebcfa2d22fbf93/python/bindings.cc#L24

**Comment by humanbeingZ** [comment]
![this](https://user-images.githubusercontent.com/25080075/155742881-ceca6102-e661-4eef-a759-8ed49b66c8bc.png)

**Comment by syoyo** [comment]
wavefront obj uses right-handed, counter-clockwise to define face orientation.

https://en.wikipedia.org/wiki/Wavefront_.obj_file

**Comment by humanbeingZ** [comment]
I'm really sorry for the bother. I think I might have misunderstood something. When I try to load the following obj (which is pure triangle), the result looks not good. Does this mean that my code above is not doing what I want? 
```
####
#
# OBJ File Generated by Meshlab
#
####
# Object pillow_tri_sim_50.obj
#
# Vertices: 27
# Faces: 50
#
####
v 103.728752 40.225372 -22.638380
v 104.858101 72.017593 -32.051025
v 104.413269 64.791084 -37.072239
v 113.326988 73.320679 -27.119675
v 89.046165 59.581371 -47.348888
v 87.077049 73.973976 -45.460411
v 109.285645 61.334530 -32.280876
v 87.432045 73.265976 -51.240524
v 105.963341 72.697556 -25.378479
v 93.500771 59.907940 -45.878441
v 91.776268 51.598618 -44.513905
v 105.190163 51.401321 -32.413227
v 84.252457 70.107338 -44.106201
v 90.943184 69.846367 -31.046146
v 79.866531 50.957314 -42.565815
v 98.849701 59.492733 -20.966267
v 101.487793 45.786160 -32.357758
v 78.648308 52.305763 -33.033482
v 106.708008 61.076149 -23.766838
v 95.731483 41.209087 -32.172668
v 88.196945 54.121414 -23.249815
v 93.747871 42.721279 -21.449389
v 98.809258 43.997261 -16.878447
v 102.801628 39.554707 -16.558514
v 75.983612 43.648041 -36.780018
v 84.384369 42.946167 -43.664497
v 76.710373 40.138397 -41.981365
# 27 vertices, 0 vertices normals

f 17 1 20
f 20 27 26
f 8 3 10
f 6 2 3
f 7 3 4
f 7 10 3
f 6 3 8
f 4 3 2
f 2 9 4
f 19 4 9
f 4 19 7
f 5 8 10
f 5 15 8
f 8 15 13
f 8 13 6
f 2 6 9
f 10 7 12
f 11 10 12
f 15 5 26
f 14 13 18
f 13 14 6
f 14 9 6
f 14 16 9
f 19 9 16
f 17 11 12
f 10 11 5
f 5 11 26
f 15 18 13
f 14 21 16
f 1 12 7
f 1 7 19
f 1 17 12
f 11 17 26
f 15 25 18
f 14 18 21
f 20 26 17
f 1 19 23
f 22 21 25
f 20 1 24
f 20 24 22
f 25 27 22
f 18 25 21
f 27 20 22
f 16 21 23
f 16 23 19
f 21 22 23
f 1 23 24
f 23 22 24
f 15 26 25
f 27 25 26
# 50 faces, 0 coords texture

# End of File
```

**Comment by humanbeingZ** [comment]
![pillow](https://user-images.githubusercontent.com/25080075/155750283-4cd21eb0-8c2b-482d-b1a2-a75e8d9e1a6e.png)

**Comment by syoyo** [comment]
@humanbeingZ I meant tinyobjloader-loaded mesh's orientation is defined in right-handed, clockwise order. So you may need to reverse the triangle vertex indices(or flip normals) depending on your target usage(e.g. `off` format output)
 
You can output .obj again then test .obj's face orientation and normal direction using Blender: https://blenderbasecamp.com/home/how-to-identify-normals-in-blender-3d/

If tinyobjloader-loaded .obj's face orientation was wrong(i.e. not clockwise in right-handed coordinate) for the mesh composed purely with triangles, it might be a tinlyobjloader's bug so please reopen the issue.

**Comment by humanbeingZ** [comment]
> I'm really sorry for the bother. I think I might have misunderstood something. When I try to load the following obj (which is pure triangle), the result looks not good. Does this mean that my code above is not doing what I want?
> 
> ```
> ####
> #
> # OBJ File Generated by Meshlab
> #
> ####
> # Object pillow_tri_sim_50.obj
> #
> # Vertices: 27
> # Faces: 50
> #
> ####
> v 103.728752 40.225372 -22.638380
> v 104.858101 72.017593 -32.051025
> v 104.413269 64.791084 -37.072239
> v 113.326988 73.320679 -27.119675
> v 89.046165 59.581371 -47.348888
> v 87.077049 73.973976 -45.460411
> v 109.285645 61.334530 -32.280876
> v 87.432045 73.265976 -51.240524
> v 105.963341 72.697556 -25.378479
> v 93.500771 59.907940 -45.878441
> v 91.776268 51.598618 -44.513905
> v 105.190163 51.401321 -32.413227
> v 84.252457 70.107338 -44.106201
> v 90.943184 69.846367 -31.046146
> v 79.866531 50.957314 -42.565815
> v 98.849701 59.492733 -20.966267
> v 101.487793 45.786160 -32.357758
> v 78.648308 52.305763 -33.033482
> v 106.708008 61.076149 -23.766838
> v 95.731483 41.209087 -32.172668
> v 88.196945 54.121414 -23.249815
> v 93.747871 42.721279 -21.449389
> v 98.809258 43.997261 -16.878447
> v 102.801628 39.554707 -16.558514
> v 75.983612 43.648041 -36.780018
> v 84.384369 42.946167 -43.664497
> v 76.710373 40.138397 -41.981365
> # 27 vertices, 0 vertices normals
> 
> f 17 1 20
> f 20 27 26
> f 8 3 10
> f 6 2 3
> f 7 3 4
> f 7 10 3
> f 6 3 8
> f 4 3 2
> f 2 9 4
> f 19 4 9
> f 4 19 7
> f 5 8 10
> f 5 15 8
> f 8 15 13
> f 8 13 6
> f 2 6 9
> f 10 7 12
> f 11 10 12
> f 15 5 26
> f 14 13 18
> f 13 14 6
> f 14 9 6
> f 14 16 9
> f 19 9 16
> f 17 11 12
> f 10 11 5
> f 5 11 26
> f 15 18 13
> f 14 21 16
> f 1 12 7
> f 1 7 19
> f 1 17 12
> f 11 17 26
> f 15 25 18
> f 14 18 21
> f 20 26 17
> f 1 19 23
> f 22 21 25
> f 20 1 24
> f 20 24 22
> f 25 27 22
> f 18 25 21
> f 27 20 22
> f 16 21 23
> f 16 23 19
> f 21 22 23
> f 1 23 24
> f 23 22 24
> f 15 26 25
> f 27 25 26
> # 50 faces, 0 coords texture
> 
> # End of File
> ```

For the aforementioned obj file, if I choose to load it in, and then save it as ***obj*** again using the following code, the face normal is still not right.
```python3
import sys
import numpy as np
import tinyobjloader


def process(obj_path):
    reader = tinyobjloader.ObjReader()
    ret = reader.ParseFromFile(obj_path)

    if not ret:
        print('Failed to load: ', obj_path)
        sys.exit(-1)
    
    attrib = reader.GetAttrib()
    vertices = np.array(attrib.vertices).reshape(-1, 3)
    shape = reader.GetShapes()[0]

    indices = shape.mesh.indices
    faces = [idx.vertex_index for idx in indices]
    faces = np.array(faces).reshape(-1, 3)

    with open(f'out.obj', 'w') as f:
        for v in vertices:
            print('v', *[f'{i:0.2f}' for i in v], sep=' ', file=f)
        for face in faces:
            print('f', *(face + 1), sep=' ', file=f)


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('Usage: python <obj_path>')
        sys.exit(0)

    obj_path = sys.argv[1]
    process(obj_path)
```

I'm not familiar with blender. However, I guess by comparing input/output obj file line by line would also do the thing?
 
![2022-02-26_11-29](https://user-images.githubusercontent.com/25080075/155827237-e2a43708-81f0-4ca8-9c6e-4d78d36bdf86.png)

**Comment by humanbeingZ** [comment]
@syoyo Really sorry for all the inconvenience that I caused. I didn't find the option to reopen the issue.

**Comment by syoyo** [comment]
@humanbeingZ Sorry for the late reply, but the issue was exist in built-in tessellator  and now fixed in this commit: https://github.com/tinyobjloader/tinyobjloader/tree/tylermorganwall-triangulation-fix

Please try that branch. At least now it can handle your model correctly.

![Screenshot from 2022-10-02 19-47-33](https://user-images.githubusercontent.com/18676/193450377-03cb47f2-0611-415b-a24a-4f0a6ad07e90.png)

**Comment by humanbeingZ** [comment]
Really thanks, syoyo. However, I tried the new branch and the problem still exists. And the face with false orientation does not come from triangulated quad face, but from original triangle face. I'm still using [this code](https://github.com/tinyobjloader/tinyobjloader/issues/330#issuecomment-1051548353). 

When I specify version using `pip install tinyobjloader==2.0.0rc8`, the problem is gone.
Otherwise, both trying out the [`tylermorganwall-triangulation-fix`](https://github.com/tinyobjloader/tinyobjloader/tree/tylermorganwall-triangulation-fix) branch and version 2.0.0rc9 does not help.

**Comment by syoyo** [comment]
You shoud test with this code https://github.com/tinyobjloader/tinyobjloader/issues/319#issue-995761553 , not yours
